### DISCUSSION TITLE
MonitorControl v4.0.0

### DISCUSSION THREAD

**Discussion by JoniVR** [MonitorControl/MonitorControl#741] [Release Discussion]
### Enhancements

- Automatic & manual updates through the app, no more manual downloads 🎉 
- Moved 'Show advanced settings' under 'Displays'
- Added proper support for controlling Apple displays.
- Added option to show/hide brightness slider.
- Added option to show brightness slider for internal display and apple displays (enabled by default).
- Replication of built-in and Apple display brightness to corresponding brightness slider.
- Added suffix to similarly named displays for better differentiation.
- Option to disable slider snapping for finer control + disable slider snapping by default.
- Added option to show slider tick marks for better accuracy.
- Added option to use window focus instead of mouse to determine which display to control.
- <kbd>control</kbd> + <kbd>command</kbd> + <kbd>brightness</kbd> now controls external displays only (<kbd>control</kbd> + <kbd>Brightness</kbd> continues to control internal display only)
- Added separate tab for menu options.
- Added option to restore last saved values upon startup.
- Added option for audio device name matching for display volume control selection.
- Separated option to change all screens for brightness and volume.
- Added option for keyboard fine scale for brightness.
- Added option for keyboard fine scale for volume.
- Added version check upon startup for mandatory preferences reset upon downgrade or incompatible previous version + notification about this.
- Added implementation for <kbd>command</kbd> + <kbd>f1</kbd> macOS shortcut to enable/disable mirroring.
- Added safer 'Assume last saved settings are valid' option as default instead of startup DDC read (or restore).
- Streamlined preference panes, 'Show advanced settings' now affect all tabs. This leads to a better and safer first timer experience (especially because of the influx of many new features).
- Added a Quit button to Preferences if menu is hidden (it was not passible to quit the application until this time in this mode only by re-enabling the menu).
- Lowered default first-run volume DDC default from 75% to 15% if read is not possible or disabled to prevent unexpectedly loud sound. 
- Added slider skew setting on a per control basis to have the ability to manipulate DDC slider balance and OSD scale if display control is not linear.
- Added the ability to set min. and max. DDC bounds on a per display, per control basis.
- Audio device name override option for a display (manually assign a specific audio device to a display).
- Advanced setting to invert DDC control range (some displays have the scale reversed).
- Advanced setting to remap DDC control code (some displays have contrast and brightness mixed up).
- Ability to mark a DDC control as available or unavailable in advanced settings under Displays.
- Ability to automatically hide menu icon if there is no slider present in the menu.
- Option to show slider percentage for more precision.
- Option to set combined or separate OSD scale when combined hardware+software brightness is used.
- Apple like smooth brightness change (both for software, hardware, mixed and DisplayServices).
- Added support for DisplayLink, AirPlay, Sidecar, screen sharing etc. using window shades (this is an inferior technique to the existing software implementation - gamma control - but still better than nothing). Disabled for any kind of mirroring setups. [^1]
- Brightness change synchronisation from Built-In and Apple displays to other displays. This makes Touch Bar, Ambient light sensor, Control Center and System Preferences induced changes affect all displays. Synchronisation uses a sophisticated indirect delta method + the user can intervene and adjust individual screen brightness at any time to easily compensate mismatching native brightness levels.
- Preferences pane tab selector has a simpler look on Catalina.
- All menu sliders are now scrollable using a magic mouse/trackpad swipes or mouse wheel.
- Added option for menu to show only items that are relevant to display which shows the menu currently.
- Added option to enable combined sliders (note: this option combined with enabled Apple/built-in display syncing and enabled 'change all' keyboard settings finally provides full synchronised control of all displays).
- Combined sliders can now display multiple displays when keyboard and brightness syncing is not enabled. [^1]
- Redesigned sliders to look like Big Sur/Monterey Control Center's sliders. [^1]
- Quit and Preferences... are now icons for a much cleaner look. [^1]
- Added option to change additional menu options style or hide them. [^1]
- Multiple displays are now in nice Big Sur styled blocks - no more ugly separators. [^1]
- Added customisable gamma/ddc switchover point for combined brightness in the advanced section of Displays.
- Added comma separated list for control code override to enable edge cases like controlling Brightness and Contrast at the same time (use VCP list entry `10, 12` for that)
- Contrast can now be controlled from keyboard via <kbd>control</kbd> + <kbd>option</kbd> + <kbd>command</kbd> + <kbd>brightness up/down</kbd>.
- Custom keyboard shortcuts for brightness, contrast, volume and mute
- Added Internet Access Policy
- Added advanced option to enable zero brightness with software dimming.
- Added <kbd>command</kbd> + <kbd>q</kbd> shortcut in menu when it is in standard text mode.
- Rearranged Preferences to fit better on smaller screens in Advanced mode
- Alert when repeated interference is detected with applications.
- Added 'Avoid gamma table manipulation' option for peaceful coexistence with f.lux.

### Other under the hood changes and bug fixes

- Standardised internal scale among various displays and DDC ranges for ranged controls.
- Uses the new internal scale for combined hardware-software brightness mode.
- Improved settings reset functionality for virtual and Apple displays.
- Migrated scales to internal float representation to prevent loss of fine detail on transformations.
- Fixed double sound when muting multiple external displays at the same time.
- Multiple attempts to write last DDC brightness and contrast settings after wake when 'Apply last saved settings' is enabled to make sure that slow displays are handled well
- Fixed lack of initial volume configuration if slider is not shown in menu.
- Fixed wrong settings being applied to a display when replaced on Apple Silicon (UserDefaults preferences are now tied to specific display strings instead of CGDirectDisplayID - which is no longer semi-unique on arm64).
- A lot of refactoring, streamlining and general optimisations.

### Thanks to all our translators

- Chinese - thanks to @dev-coco, @anmoliyang
- Chinese (Traditional, Taiwan) - thanks to @stiiveo, @dororojames, @kix99aug
- Dutch - thanks to @JoniVR 
- French - thanks to @the0neyouseek 
- German - thanks to @jajoho, @curana
- Hungarian - thanks to @waydabber
- Italian - thanks to @picov
- Korean - thanks to @zzulu
- Turkish - thanks to @mennan, @ozkanozcan

### Notes

- If you discover issues, feel free to open an issue!
- If you want to help translating, see #637
- Special thanks to @waydabber on another fantastic job on this release!
- Thanks to everyone who helps make MonitorControl better!

[^1]: Only on Big Sur and above

<hr /><em>This discussion was created from the release <a href='https://github.com/MonitorControl/MonitorControl/releases/tag/v4.0.0'>MonitorControl v4.0.0</a>.</em>

**Comment by waydabber** [comment]
For those who are wondering, here are the changes compared to the RC. :)

### Enhancements:

- Added 'Check for Updates' button under 'About'
- Moved 'Show advanced settings' under 'Displays'
- Software brightness changes are reapplied upon startup when 'Apply last saved values to the display' is selected under 'Upon startup or wake'.
- Multiple attempts to write last DDC brightness and contrast settings after wake when 'Apply last saved settings' is enabled to make sure that slow displays are handled well
- Small change of wording under Displays to reflect the fact that software dimming is not always gamma now
- Improved settings reset functionality for virtual and Apple displays.
- Brightness control availability can be toggled for built-in and virtual displays as well.
- Updated localizations.

### Fixed issues:

- On some rare occasions, macOS does not reset gamma tables upon display reconfiguration which caused an erroneus baseline and double dimming in software and combined mode.
- Virtual displays were not ignored in gamma interference detection causing unneeded gamma interference popup messages.
- Too long display names looks bad in the menu when multiple displays are present.
- Fixed internal errors about unneeded missing outlet connections.
- Menu shows blank space for display with no controls.

### Under the hood changes

- Some minor tweaks regarding how software dimming and gamma table manipulation behaves.
- Enabled Javascript for Sparkle release notes
- Log levels are now standardized
- Removed localizations that were not updated for 4.0.0 (will add them back once they are completed)

**Reply by alexandrosraikos** [reply]
This is excellent work! 💯

**Comment by jajoho** [comment]
Thanks a lot!

In one place the text of the German translation is cut off: Under "App Menu > multiple displays" the description is not displayed completely if "multiple displays" is selected.

<img width="842" alt="Bildschirmfoto 2021-10-26 um 21 45 52" src="https://user-images.githubusercontent.com/15175599/138950278-4583a08f-47bd-4149-b53f-894b3bd5a3ae.png">

**Reply by waydabber** [reply]
Hi @jajoho, yes, I see. This line is set to truncate if it is too long. I can easily change it to wrap instead (that will make the window bigger). The other solution is to shorten the text a bit. Which one would you like?

We plan a 4.0.1 service release for issues that come up in 4.0.0.

**Reply by jajoho** [reply]
Hi @waydabber,

I do not have Xcode on my Mac. So I don't know if it will work with this shortened text:
"Funktioniert mit variabler Synchronisierung und 'Steuerung aller'-Tastatureinstellung am besten"

**Reply by waydabber** [reply]
Well, almost, but still does not fit one line. :) I can enable wraps if that helps and it cannot be shortened anymore.

<img width="504" alt="Screen Shot 2021-10-26 at 22 25 30" src="https://user-images.githubusercontent.com/37590873/138955602-b3e293e5-4360-41c4-b018-1fb4172db0a4.png">

**Reply by jajoho** [reply]
In this case I think wrap would be great. I also have found a missing string in the German translation and a typo in this string (the original translation, not the shorter version). Do you want me to send a pull request for those?

**Reply by waydabber** [reply]
Yes, please create a PR and also a bug issue for the wrap so I can fix it tomorrow (btw - I found other ones that should be wraps instead of truncate for the sake of consistency).

Thank you! :)

**Reply by jajoho** [reply]
created issue #743 (and pr #742)

Thank you very much!

**Comment by Lewis-Xie** [comment]
I've been watching for an app like this for so long, it's working on my M1 Mac mini perfectly. Thank you so much!

**Reply by arousta** [reply]
I want to say just thank you. I tried couple other stuff and this one only worked. Also very nice UI and integration with mac launcher. Thanks! 👏 👏 👏
